test/l1/L1DAIWormholeBridge.ts

```diff
@@ -0,0 +1,129 @@
+import { assertPublicMutableMethods, simpleDeploy, waitForTx } from '@makerdao/hardhat-utils'
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/dist/src/signer-with-address'
+import { expect } from 'chai'
+import { ethers } from 'hardhat'
+
+import { Dai__factory, L1DAIWormholeBridge__factory, L1Escrow__factory } from '../../typechain-types'
+import { deployAbstractMock, deployMock, deployOptimismContractMock } from '../helpers'
+
+const INITIAL_ESCROW_BALANCE = 3000
+const SOURCE_DOMAIN_NAME = ethers.utils.formatBytes32String('optimism-a')
+const TARGET_DOMAIN_NAME = ethers.utils.formatBytes32String('arbitrum-a')
+const AMOUNT = 100
+
+describe('L1DAIWormholeBridge', () => {
+  it('has correct public interface', async () => {
+    await assertPublicMutableMethods('L1DAIWormholeBridge', [
+      'finalizeFlush(bytes32,uint256)',
+      'finalizeRegisterWormhole((bytes32,bytes32,address,address,uint128,uint80,uint48))',
+    ])
+  })
+
+  describe('constructor', () => {
+    it('assigns all variables properly', async () => {
+      const [l2DAIWormholeBridge, l1CrossDomainMessenger, l1Escrow, wormholeRouter] = await ethers.getSigners()
+
+      const l1Dai = await simpleDeploy<Dai__factory>('Dai', [])
+      const l1DAITokenBridge = await simpleDeploy<L1DAIWormholeBridge__factory>('L1DAIWormholeBridge', [
+        l1Dai.address,
+        l2DAIWormholeBridge.address,
+        l1CrossDomainMessenger.address,
+        l1Escrow.address,
+        wormholeRouter.address,
+      ])
+
+      expect(await l1DAITokenBridge.l1Token()).to.eq(l1Dai.address)
+      expect(await l1DAITokenBridge.l2DAIWormholeBridge()).to.eq(l2DAIWormholeBridge.address)
+      expect(await l1DAITokenBridge.escrow()).to.eq(l1Escrow.address)
+      expect(await l1DAITokenBridge.messenger()).to.eq(l1CrossDomainMessenger.address)
+      expect(await l1DAITokenBridge.wormholeRouter()).to.eq(wormholeRouter.address)
+    })
+  })
+
+  describe('finalizeFlush', () => {
+    it('calls the router to settle the dai debt', async () => {
+      const [_, l1MessengerImpersonator, routerImpersonator, wormholeJoin] = await ethers.getSigners()
+      const {
+        l1Dai,
+        l1DAIWormholeBridge,
+        l1CrossDomainMessengerMock,
+        l2DAIWormholeBridge,
+        l1Escrow,
+        wormholeRouterMock,
+      } = await setupTest({ l1MessengerImpersonator, routerImpersonator })
+      l1CrossDomainMessengerMock.smocked.xDomainMessageSender.will.return.with(() => l2DAIWormholeBridge.address)
+
+      await waitForTx(l1DAIWormholeBridge.connect(l1MessengerImpersonator).finalizeFlush(TARGET_DOMAIN_NAME, AMOUNT))
+      const routerSettleCallData = wormholeRouterMock.smocked.settle.calls[0]
+      await waitForTx(
```

Comment:
IMO this is unneeded from POV of a unit test. We should simply test that finalizeFlush moved DAI to the bridge b/c this is what router expects.

Reply:
This allows to check that:
(1) the approval to the router was done correctly.
(2) the DAI moved from the escrow to some new address.

We could check (2) by simply checking that the balance of the escrow has been reduced. IMO checking that the bridge balance has increased would be a bit confusing to the reader as it's not how the bridge works outside unit tests.

But I don't see another way to check (1) other than mocking a transfer of DAI to the router. Ideally this would be done in a `wormholeRouterMock.smocked.settle.will.return.with(...)` hook but unfortunately this hook doesn't accept async function.

Reply:
Thinking about this more, I think it's fine to leave check (1) to integration tests and reduce the scope of this unit test

Reply:
I think about it like this: units should test expected behavior of a particular unit. In this case, it is expected that l1WormholeBridge will move `amt` tokens from the escrow to itself and call the router.